src/javascript/app/base/network_monitor.js

```diff
@@ -0,0 +1,118 @@
+const Header              = require('./header');
+const BinarySocket        = require('./socket');
+const BinarySocketGeneral = require('./socket_general');
+const getElementById      = require('../../_common/common_functions').getElementById;
+const localize            = require('../../_common/localize').localize;
+
+/*
+ * Monitors the network status and initialises the WebSocket connection
+ * 1. online : check the WS status (init/send: blink after timeout, open/message: online)
+ * 2. offline: it is offline
+ */
+const NetworkMonitor = (() => {
+    const status_config = {
+        online  : { class: 'online',  tooltip: localize('Online') },
+        offline : { class: 'offline', tooltip: localize('Offline') },
+        blinking: { class: 'pulser',  tooltip: localize('Connecting') },
+    };
+    const pendings = {};
+    const pending_keys = {
+        ws_init   : 'ws_init',
+        ws_request: 'ws_request',
+    };
+    const pending_timeouts = {
+        [pending_keys.ws_init]   : 5000,
+        [pending_keys.ws_request]: 10000,
+    };
+
+    let ws_config,
+        el_status,
+        network_status;
+
+    const init = () => {
+        ws_config = $.extend({ wsEvent }, BinarySocketGeneral.initOptions());
+        el_status = getElementById('network_status');
+
+        if ('onLine' in navigator) {
+            window.addEventListener('online',  setStatus);
+            window.addEventListener('offline', setStatus);
+        } else { // if not supported, default to online and fallback to WS checks
+            navigator.onLine = true;
+        }
+
+        if (isOnline()) {
+            BinarySocket.init(ws_config);
+        }
+
+        setStatus(isOnline() ? 'online' : 'offline');
+    };
+
+    const isOnline = () => navigator.onLine;
+
+    const wsReconnect = () => {
+        if (isOnline() && BinarySocket.hasReadyState(2, 3)) { // CLOSING or CLOSED
+            BinarySocket.init(ws_config);
+            clearPendings();
+        } else {
+            BinarySocket.send({ ping: 1 }); // trigger a request to get stable status sooner
+        }
+    };
+
+    const setStatus = (status) => {
+        if (!isOnline() || status === pending_keys.ws_init) {
```

Comment:
The `status` field is slightly misleading here, I think: `ws_init` doesn't really mean "we're in `ws_init state", it's specifically "we had a timeout in `ws_init` state"? At the moment, the logic for this block could be read as "if we've just started the connection, display the 'Connection error' text" (whereas that doesn't actually happen until after the timeout has expired).

I'd suggest marking this as blinking while we're waiting for the initial websocket connection, by the way - seems more accurate than claiming we're connected?

Reply:
You're right, it's misleading. Will change it.
Thanks.